ecosystem/indexer-grpc/indexer-grpc-table-info/src/table_info_parser.rs

```diff
@@ -0,0 +1,332 @@
+        loop {
+            let start_time = std::time::Instant::now();
+            let ledger_version = self.get_highest_known_version().await.unwrap_or_default();
+            let batches = self.get_batches(ledger_version).await;
+            let results = self
+                .process_multiple_batches(db.clone(), batches, ledger_version)
+                .await;
+            let max_version = self.get_max_batch_version(results).unwrap_or_default();
+            let versions_processed = max_version - self.current_version + 1;
+
+            log_grpc_step(
+                SERVICE_TYPE,
+                IndexerGrpcStep::TableInfoProcessed,
+                self.enable_verbose_logging,
+                Some(self.current_version as i64),
+                Some(max_version as i64),
+                None,
+                None,
+                Some(start_time.elapsed().as_secs_f64()),
+                None,
+                Some(versions_processed as i64),
+                None,
+            );
+
+            self.current_version = max_version + 1;
+        }
+    }
+
+    /// Fans out a bunch of threads and processes write sets from transactions in parallel.
+    /// Pushes results in parallel to the stream, but only return that the batch is
+    /// fully completed if every job in the batch is successful and no pending on items
+    /// Processing transactions in 2 stages:
+    /// 1. Fetch transactions from ledger db
+    /// 2. Get write sets from transactions and parse write sets to get handle -> key,value type mapping, write the mapping to the rocksdb
+    async fn process_multiple_batches(
+        &self,
+        db: DbReaderWriter,
+        batches: Vec<TransactionBatchInfo>,
+        ledger_version: u64,
+    ) -> Vec<Result<EndVersion, Status>> {
+        let mut tasks = vec![];
+        let db_writer = db.writer.clone();
+        let context = self.context.clone();
+
+        for batch in batches.iter().cloned() {
+            let task = tokio::spawn(Self::process_single_batch(
+                context.clone(),
+                db_writer.clone(),
+                ledger_version,
+                batch,
+                false, /* end_early_if_pending_on_empty */
+                self.enable_verbose_logging,
+            ));
+            tasks.push(task);
+        }
+
+        match futures::future::try_join_all(tasks).await {
+            Ok(res) => {
+                let last_batch = batches.last().cloned().unwrap();
+                let total_txns_to_process = last_batch.start_version
+                    + last_batch.num_transactions_to_fetch as u64
+                    - self.current_version;
+                let end_version =
+                    last_batch.start_version + last_batch.num_transactions_to_fetch as u64;
+
+                // Update rocksdb next version after threads join
+                db.writer
+                    .clone()
+                    .update_next_version(end_version + 1)
+                    .unwrap();
```

Comment:
Oh we shouldn't update this here. We should update this at the end of the pending_on thing otherwise if it crashes it'll skip.

Reply:
good catch!!!